md5sum_cli.cpp

```diff
@@ -0,0 +1,268 @@
+            case QUIET_OPTION:
+                quiet = true;
+                break;
+            case STATUS_OPTION:
+                status_only = true;
+                break;
+            case 't':
+                have_input_mode_option = true;
+                break;
+            case HELP_OPTION:
+                show_help = true;
+                break;
+            case VERSION_OPTION:
+                show_version = true;
+                break;
+            default:
+                // Error message was already given from getopt() code, so no need
+                // to output anything else here, but do flush its output so that it
+                // appears before the usage message.
+                std::fflush(stderr);
+
+                command_line_syntax_error = true;
+            }
+
+        if(command_line_syntax_error)
+            break;
+        }
+
+    if(command_line_syntax_error)
+        {
+        std::cerr << "Try 'md5sum --help' for more information." << std::endl;
+        return EXIT_FAILURE;
+        }
+    if(have_input_mode_option && do_check)
+        {
+        std::cerr
+            << "The --binary and --text options are meaningless when "
+            << "verifying checksums." << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(status_only && !do_check)
+        {
+        std::cerr
+            << "The --status option is meaningful only when verifying checksums."
+            << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(quiet && !do_check)
+        {
+        std::cerr
+            << "The --quiet option is meaningful only when verifying checksums."
+            << std::endl;
+        return usage(EXIT_FAILURE);
+        }
+    if(show_help)
+        {
+        return usage(EXIT_SUCCESS);
+        }
+
+    if(show_version)
+        {
+        version();
+        return EXIT_SUCCESS;
+        }
+
+    for(int i = getopt_long.optind; i < argc; ++i)
+        {
+        char const* filename = argv[i];
+        std::string error_message;
+        std::string md5;
+
+        if(do_check)
+            {
+            std::vector<md5sum_for_file> sums;
+
+            try
+                {
+                sums = md5sum_vec_from_file(filename);
+                }
+            catch(std::runtime_error const& e)
+                {
+                if(!status_only)
+                    std::cerr << "md5sum: " << e.what() << std::endl;
+                return EXIT_FAILURE;
+                }
+
+            for(auto const& s : sums)
+                {
+                try
+                    {
+                    md5 = calculate_md5sum(s.filename);
+                    }
+                catch(std::runtime_error const& e)
+                    {
+                    if(!status_only)
+                        std::cerr << "md5sum: " << e.what() << std::endl;
+                    return EXIT_FAILURE;
+                    }
+
+                if(md5 != s.md5sum)
```

Comment:
I'd write this entire block as:
```cpp
bool const current_ok = md5 == s.md5sum;
if(!status_only)
    {
    if(!current_ok || !quiet)
        {
        std::cout << s.filename << ": " << (current_ok ? "OK" : "FAILED") << std::endl;
        }
    }
ok |= current_ok;
```
to avoid duplicating `status_only` checks and the actual output.

I'd also consider renaming `ok` to `all_ok`.